fastestimator/trace/trace.py

```diff
@@ -47,26 +47,21 @@ def __init__(self,
     def on_begin(self):
         """Runs once at the beginning of training
         """
-
     def on_epoch_begin(self):
         """Runs at the beginning of each epoch
         """
-
     def on_batch_begin(self):
```

Comment:
1) trace might take inputs generated by other traces
       - Lack of output spec prevents trace sorting / true warmup validation
       - This would prevent get_inputs_by_key from working
2)  trace may want to modify batch data, impossible since no data in on_batch_begin
       - RL functionality will probably rely on this
3) Should system be passed as function arg rather than being an instance variable? alternatively, should we make it a class variable rather than an instance variable? Potential problem could then occur in jupiter if user has multiple estimators at once though

Reply:
1. Trace won't write data to batch data, but instead Trace will use self.system.buffer to communicate

2. We won't allow trace to modify batch data, because there's no way to do warm up on it.  We don't have to worry about RL now.

3. I didn't get the issue. each estimator instance will have their own system instance, just like each model instance has their own variables and optimizer.

Reply:
1. many of the current traces (tensor board, XAI as examples) link outputs from one trace to inputs of another. current system allows validation of that linking but in new system with no trace outputs it's not possible to validate

2. fine, though I anticipate that we will regret this decision in the future. we might want to put a placeholder variable in the method signature so that we can implement something later without invalidating all user traces

3. technically each trace doesn't have their own system but all references to a single system instance. I'm just worried about so many references sitting around for the whole training time but I guess it's fine

Reply:
yeah I agree with your point1, we can discuss how to better approach it after merging this PR :)